With linked table to postgreSQL, case-sensivite column names not respected in where part

Hi, (Thank for your efforts to maintain this great tool)

We have a problem with a linked table to a PostgreSQL database which contains some case sensitive columns. Its look like, if we make a select on the linked table the column are not correctly send in the "where", the double quotes are removed.

We use tha H2 version 2.1.210
and the PostgreSQL version 14.1, compiled by Visual C++ build 1914, 64-bit

What ever H2 compatibility mode used (STRICT or POSTGRESQL) the error raises.

Here is an example to reproduce this problem, in the PostgrSQL database there this table:

```
CREATE TABLE "Source"."Employees" (
"EmployeeId" int4 NULL,
"EmployeeName" varchar NULL,
"CITIES" varchar NULL,
country varchar NULL
);
```
In the H2 database we create the corresponding linked table like this:

```
CREATE GLOBAL TEMPORARY LINKED TABLE IF NOT EXISTS
   "Source".LT('org.postgresql.Driver','jdbc:postgresql://localhost:5432/extract','postgres','pwd','"Source"','"Source"."Employees"')
```

Can not query the linked table with a clause WHERE on mixed case columns, H2 rebuilds the query and removes the double quotes whatever the H2 compatibility mode:

```
select "EmployeeName" from "Source".LT where "EmployeeName" = 'Nina' 
```

throw the following error:

org.h2.jdbc.JdbcSQLNonTransientException: Error accessing linked table with SQL statement "SELECT * FROM ""Source"".""Employees"" T WHERE EmployeeName>=? AND EmployeeName<=?", cause: "org.postgresql.util.PSQLException: ERREUR: la colonne « employeename » n'existe pas\000a Indice\00a0: Peut-être que vous souhaitiez référencer la colonne « t.EmployeeName ».\000a Position\00a0: 44"; 
SQL statement:
select * from "Source".LT where "EmployeeName" = 'Nina' [90111-210]
at org.h2.message.DbException.getJdbcSQLException(DbException.java:573)
at org.h2.message.DbException.getJdbcSQLException(DbException.java:496)
at org.h2.message.DbException.get(DbException.java:216)
at org.h2.table.TableLink.wrapException(TableLink.java:495)
at org.h2.index.LinkedIndex.find(LinkedIndex.java:132)
at org.h2.index.IndexCursor.find(IndexCursor.java:161)
at org.h2.table.TableFilter.next(TableFilter.java:394)
at org.h2.command.query.Select$LazyResultQueryFlat.fetchNextRow(Select.java:1825)
at org.h2.result.LazyResult.hasNext(LazyResult.java:78)
at org.h2.result.FetchedResult.next(FetchedResult.java:34)
at org.h2.command.query.Select.queryFlat(Select.java:728)
at org.h2.command.query.Select.queryWithoutCache(Select.java:833)
at org.h2.command.query.Query.queryWithoutCacheLazyCheck(Query.java:197)
at org.h2.command.query.Query.query(Query.java:494)
at org.h2.command.query.Query.query(Query.java:457)
at org.h2.command.CommandContainer.query(CommandContainer.java:256)
at org.h2.command.Command.executeQuery(Command.java:190)
at org.h2.server.TcpServerThread.process(TcpServerThread.java:353)
at org.h2.server.TcpServerThread.run(TcpServerThread.java:191)
at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: org.postgresql.util.PSQLException: ERREUR: la colonne « employeename » n'existe pas
Indice : Peut-être que vous souhaitiez référencer la colonne « t.EmployeeName ».
Position : 44
at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2553)
at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2285)
at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:323)
at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:481)
at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:401)
at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:164)
at org.postgresql.jdbc.PgPreparedStatement.execute(PgPreparedStatement.java:153)
at org.h2.table.TableLink.execute(TableLink.java:552)
at org.h2.index.LinkedIndex.find(LinkedIndex.java:128)
... 15 more
at org.h2.message.DbException.getJdbcSQLException(DbException.java:573)
at org.h2.engine.SessionRemote.readException(SessionRemote.java:637)
at org.h2.engine.SessionRemote.done(SessionRemote.java:606)
at org.h2.command.CommandRemote.executeQuery(CommandRemote.java:171)
at org.h2.jdbc.JdbcStatement.executeInternal(JdbcStatement.java:247)
at org.h2.jdbc.JdbcStatement.execute(JdbcStatement.java:223)
at org.h2.server.web.WebApp.getResult(WebApp.java:1339)
at org.h2.server.web.WebApp.query(WebApp.java:1137)
at org.h2.server.web.WebApp$1.next(WebApp.java:1103)
at org.h2.server.web.WebApp$1.next(WebApp.java:1090)
at org.h2.server.web.WebThread.process(WebThread.java:189)
at org.h2.server.web.WebThread.run(WebThread.java:102)
at java.base/java.lang.Thread.run(Thread.java:833)

--- a/h2/src/docsrc/html/changelog.html
+++ b/h2/src/docsrc/html/changelog.html
@@ -21,6 +21,8 @@ Change Log
 
 <h2>Next Version (unreleased)</h2>
 <ul>
+<li>Issue #3448: With linked table to postgreSQL, case-sensitive column names not respected in where part
+</li>
 <li>Issue #3434: JavaTableFunction is not closing underlying ResultSet when reading column list
 </li>
 <li>Issue #3468: Invalid DB format exception (for 1.x DB in 2.x h2) should have a specific SQLException vendorCode

--- a/h2/src/docsrc/html/changelog.html
+++ b/h2/src/docsrc/html/changelog.html
@@ -21,6 +21,8 @@ Change Log
 
 <h2>Next Version (unreleased)</h2>
 <ul>
+<li>Issue #3434: JavaTableFunction is not closing underlying ResultSet when reading column list
+</li>
 <li>Issue #3468: Invalid DB format exception (for 1.x DB in 2.x h2) should have a specific SQLException vendorCode
 </li>
 <li>Issue #3528: Weird syntax error with HAVING clause in Oracle Mode

--- a/h2/src/docsrc/html/changelog.html
+++ b/h2/src/docsrc/html/changelog.html
@@ -21,6 +21,8 @@ Change Log
 
 <h2>Next Version (unreleased)</h2>
 <ul>
+<li>Issue #3468: Invalid DB format exception (for 1.x DB in 2.x h2) should have a specific SQLException vendorCode
+</li>
 <li>Issue #3528: Weird syntax error with HAVING clause in Oracle Mode
 </li>
 <li>Issue #3307: Fix SHUTDOWN DEFRAG for encrypted databases

--- a/h2/src/main/org/h2/index/LinkedIndex.java
+++ b/h2/src/main/org/h2/index/LinkedIndex.java
@@ -97,7 +97,7 @@ public class LinkedIndex extends Index {
                 builder.append(f ? " AND " : " WHERE ");
                 f = true;
                 Column col = table.getColumn(i);
-                col.getSQL(builder, sqlFlags);
+                addColumnName(builder, col);
                 if (v == ValueNull.INSTANCE) {
                     builder.append(" IS NULL");
                 } else {
@@ -113,7 +113,7 @@ public class LinkedIndex extends Index {
                 builder.append(f ? " AND " : " WHERE ");
                 f = true;
                 Column col = table.getColumn(i);
-                col.getSQL(builder, sqlFlags);
+                addColumnName(builder, col);
                 if (v == ValueNull.INSTANCE) {
                     builder.append(" IS NULL");
                 } else {
@@ -130,6 +130,36 @@ public class LinkedIndex extends Index {
             return new LinkedCursor(link, rs, session, sql, prep);
         } catch (Exception e) {
             throw TableLink.wrapException(sql, e);
+        }
+    }
+
+    private void addColumnName(StringBuilder builder, Column col) {
+        String identifierQuoteString = link.getIdentifierQuoteString();
+        String name = col.getName();
+        if (identifierQuoteString == null || identifierQuoteString.isEmpty() || identifierQuoteString.equals(" ")) {
+            builder.append(name);
+        } else if (identifierQuoteString.equals("\"")) {
+            /*
+             * StringUtils.quoteIdentifier() can produce Unicode identifiers,
+             * but target DBMS isn't required to support them
+             */
+            builder.append('"');
+            int i = name.indexOf('"');
+            if (i < 0) {
+                builder.append(name);
+            } else {
+                builder.append(name, 0, ++i).append('"');
+                for (int l = name.length(); i < l; i++) {
+                    char c = name.charAt(i);
+                    if (c == '"') {
+                        builder.append('"');
+                    }
+                    builder.append(c);
+                }
+            }
+            builder.append('"');
+        } else {
+            builder.append(identifierQuoteString).append(name).append(identifierQuoteString);
         }
     }
 
@@ -183,7 +213,7 @@ public class LinkedIndex extends Index {
                 builder.append("AND ");
             }
             Column col = table.getColumn(i);
-            col.getSQL(builder, sqlFlags);
+            addColumnName(builder, col);
             Value v = row.getValue(i);
             if (isNull(v)) {
                 builder.append(" IS NULL ");
@@ -235,7 +265,7 @@ public class LinkedIndex extends Index {
             if (i > 0) {
                 builder.append(" AND ");
             }
-            col.getSQL(builder, sqlFlags);
+            addColumnName(builder, col);
             Value v = oldRow.getValue(i);
             if (isNull(v)) {
                 builder.append(" IS NULL");

--- a/h2/src/main/org/h2/mvstore/db/Store.java
+++ b/h2/src/main/org/h2/mvstore/db/Store.java
@@ -92,6 +92,7 @@ public final class Store {
         boolean encrypted = false;
         if (dbPath != null) {
             String fileName = dbPath + Constants.SUFFIX_MV_FILE;
+            this.fileName = fileName;
             MVStoreTool.compactCleanUp(fileName);
             builder.fileName(fileName);
             builder.pageSplitSize(db.getPageSize());
@@ -126,12 +127,12 @@ public final class Store {
             // otherwise background thread would compete for store lock
             // with maps opening procedure
             builder.autoCommitDisabled();
+        } else {
+            fileName = null;
         }
         this.encrypted = encrypted;
         try {
             this.mvStore = builder.open();
-            FileStore fs = mvStore.getFileStore();
-            fileName = fs != null ? fs.getFileName() : null;
             if (!db.getSettings().reuseSpace) {
                 mvStore.setReuseSpace(false);
             }
@@ -155,6 +156,8 @@ public final class Store {
         switch (e.getErrorCode()) {
         case DataUtils.ERROR_CLOSED:
             throw DbException.get(ErrorCode.DATABASE_IS_CLOSED, e, fileName);
+        case DataUtils.ERROR_UNSUPPORTED_FORMAT:
+            throw DbException.get(ErrorCode.FILE_VERSION_ERROR_1, e, fileName);
         case DataUtils.ERROR_FILE_CORRUPT:
             if (encrypted) {
                 throw DbException.get(ErrorCode.FILE_ENCRYPTION_ERROR_1, e, fileName);

--- a/h2/src/main/org/h2/schema/FunctionAlias.java
+++ b/h2/src/main/org/h2/schema/FunctionAlias.java
@@ -374,13 +374,13 @@ public final class FunctionAlias extends UserDefinedFunction {
          * Create a result for the given result set.
          *
          * @param session the session
-         * @param rs the result set
+         * @param resultSet the result set
          * @param maxrows the maximum number of rows to read (0 to just read the
          *            meta data)
          * @return the value
          */
-        public static ResultInterface resultSetToResult(SessionLocal session, ResultSet rs, int maxrows) {
-            try {
+        public static ResultInterface resultSetToResult(SessionLocal session, ResultSet resultSet, int maxrows) {
+            try (ResultSet rs = resultSet) {
                 ResultSetMetaData meta = rs.getMetaData();
                 int columnCount = meta.getColumnCount();
                 Expression[] columns = new Expression[columnCount];

--- a/h2/src/main/org/h2/table/TableLink.java
+++ b/h2/src/main/org/h2/table/TableLink.java
@@ -62,6 +62,7 @@ public class TableLink extends Table {
     private boolean storesMixedCase;
     private boolean storesMixedCaseQuoted;
     private boolean supportsMixedCaseIdentifiers;
+    private String identifierQuoteString;
     private boolean globalTemporary;
     private boolean readOnly;
     private final boolean targetsMySql;
@@ -125,6 +126,7 @@ public class TableLink extends Table {
         storesMixedCase = meta.storesMixedCaseIdentifiers();
         storesMixedCaseQuoted = meta.storesMixedCaseQuotedIdentifiers();
         supportsMixedCaseIdentifiers = meta.supportsMixedCaseIdentifiers();
+        identifierQuoteString = meta.getIdentifierQuoteString();
         ArrayList<Column> columnList = Utils.newSmallArrayList();
         HashMap<String, Column> columnMap = new HashMap<>();
         String schema = null;
@@ -735,6 +737,15 @@ public class TableLink extends Table {
      */
     public int getFetchSize() {
         return fetchSize;
+    }
+
+    /**
+     * Returns the identifier quote string or space.
+     *
+     * @return the identifier quote string or space
+     */
+    public String getIdentifierQuoteString() {
+        return identifierQuoteString;
     }
 
 }

--- a/h2/src/test/org/h2/test/db/TestFunctions.java
+++ b/h2/src/test/org/h2/test/db/TestFunctions.java
@@ -44,6 +44,7 @@ import java.util.Currency;
 import java.util.Date;
 import java.util.GregorianCalendar;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.Locale;
 import java.util.Properties;
 import java.util.TimeZone;
@@ -77,6 +78,8 @@ import org.h2.value.ValueTimestampTimeZone;
 public class TestFunctions extends TestDb implements AggregateFunction {
 
     static int count;
+
+    private static HashSet<SimpleResultSet> RESULT_SETS = new HashSet<>();
 
     /**
      * Run just this test.
@@ -157,12 +160,21 @@ public class TestFunctions extends TestDb implements AggregateFunction {
     private void testFunctionTable() throws SQLException {
         Connection conn = getConnection("functions");
         Statement stat = conn.createStatement();
-        stat.execute("create alias simple_function_table for '" +
-                TestFunctions.class.getName() + ".simpleFunctionTable'");
+        synchronized (RESULT_SETS) {
+            try {
+                stat.execute("create alias simple_function_table for '" +
+                        TestFunctions.class.getName() + ".simpleFunctionTable'");
+                stat.execute("select * from simple_function_table() " +
+                        "where a>0 and b in ('x', 'y')");
+                for (SimpleResultSet rs : RESULT_SETS) {
+                    assertTrue(rs.isClosed());
+                }
+            } finally {
+                RESULT_SETS.clear();
+            }
+        }
         stat.execute("create alias function_table_with_parameter for '" +
                 TestFunctions.class.getName() + ".functionTableWithParameter'");
-        stat.execute("select * from simple_function_table() " +
-                "where a>0 and b in ('x', 'y')");
         PreparedStatement prep = conn.prepareStatement("call function_table_with_parameter(?)");
         prep.setInt(1, 10);
         ResultSet rs = prep.executeQuery();
@@ -197,6 +209,8 @@ public class TestFunctions extends TestDb implements AggregateFunction {
         result.addColumn("A", Types.INTEGER, 0, 0);
         result.addColumn("B", Types.CHAR, 0, 0);
         result.addRow(42, 'X');
+        result.setAutoClose(false);
+        RESULT_SETS.add(result);
         return result;
     }
 

--- a/h2/src/test/org/h2/test/db/TestLinkedTable.java
+++ b/h2/src/test/org/h2/test/db/TestLinkedTable.java
@@ -54,6 +54,7 @@ public class TestLinkedTable extends TestDb {
         testGeometry();
         testFetchSize();
         testFetchSizeWithAutoCommit();
+        testQuotedIdentifiers();
         deleteDb("linkedTable");
     }
 
@@ -770,6 +771,30 @@ public class TestLinkedTable extends TestDb {
         assertEquals("CREATE FORCE LINKED TABLE \"PUBLIC\".\"T\"(NULL, 'jdbc:h2:mem:one', 'sa', 'sa', 'TEST')"
                 + " FETCH_SIZE 10 AUTOCOMMIT OFF /*--hide--*/", res.getString(1));
         sb.execute("DROP TABLE T");
+        ca.close();
+        cb.close();
+    }
+
+    private void testQuotedIdentifiers() throws SQLException {
+        if (config.memory) {
+            return;
+        }
+        org.h2.Driver.load();
+        Connection ca = DriverManager.getConnection("jdbc:h2:mem:one", "sa", "sa");
+        Connection cb = DriverManager.getConnection("jdbc:h2:mem:two", "sa", "sa");
+        Statement sa = ca.createStatement();
+        Statement sb = cb.createStatement();
+        sa.execute("CREATE TABLE \"Test\" AS SELECT X \"Num\", X \"Column \"\"1\"\"\" FROM SYSTEM_RANGE(1, 100)");
+        sb.execute("CREATE LINKED TABLE T(NULL, 'jdbc:h2:mem:one', 'sa', 'sa', '\"Test\"')");
+        try (ResultSet rs = sb.executeQuery("SELECT SUM(\"Num\") FROM T WHERE \"Num\" > 1")) {
+            assertTrue(rs.next());
+            assertEquals(5049, rs.getInt(1));
+        }
+        try (ResultSet rs = sb.executeQuery(
+                "SELECT SUM(\"Column \"\"1\"\"\") FROM T WHERE \"Column \"\"1\"\"\" > 1")) {
+            assertTrue(rs.next());
+            assertEquals(5049, rs.getInt(1));
+        }
         ca.close();
         cb.close();
     }

